component/discovery/discovery.go

```diff
@@ -0,0 +1,67 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/agent/component/metrics/scrape"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const MaxUpdateFrequency = 5 * time.Second
+
+// RunDiscovery is a utility for consuming and forwarding target groups from a discoverer.
+// It will handle collating targets (and clearing), as well as time based throttling of updates.
+func RunDiscovery(ctx context.Context, d discovery.Discoverer, f func([]scrape.Target)) {
```

Comment:
Maybe leave a comment on what we expect `f()` to be, so that people won't have to go back to where it's being used?